data/blog/perfectly-fitting-text-to-container-in-react.md

```diff
@@ -0,0 +1,295 @@
+
+## The Solution Space
+
+As far as I could figure out, there are 4 and a half viable approaches.
+
+### Approach 1: SVGs
+
+SVGs are naturally scalable, that's what the "S" stands for. I figured a combination of `viewBox` trickery and with `object-fit: contain;` would do the trick, since `<svg>` supports a `<text>` element. `<text>` elements are selectable, and hopefully a screen reader can announce them? This appeals, but linearly scaling text and choosing the correct font size aren't the same.
+
+Rubik's [type tester page on Google Fonts](https://fonts.google.com/specimen/Rubik/tester) illustrates the point. Try changing the font size and you'll see how the font subtly changes characteristics. Type designers are very careful about this, especially at small sizes where legibility is a problem. Scaling an SVG does not do this, so the fonts would not look right at any size except the natural size.
+
+Aside: There's a tricky related effect where the larger a number is, the tighter the [tracking](https://en.wikipedia.org/wiki/Letter_spacing) has to be, but we'll pretend this isn't a problem for now.
+
+SVGs are a no-go, in a "pixels matter" way.
+
+### Approach 2: CSS Transforms
+
+A second possibility is using CSS transforms. `transform: scale(??)` can scale the text! This would be more complicated because we'd have to calculate the size in JavaScript and then apply it in CSS (yucky), but it also suffers from the same font scaling problem as SVGs. This is also a no-go.
+
+### Approach 3: Container Queries
+
+[CSS container queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries) are pretty widely supported now, and the premise seems helpful: scale the font size according to the container. This doesn't work either because font size controls the _height_ of the font. Scaling the font height with respect to the container Y axis height is easy, but we need to _constrain_ the size based on _both axes_. Plus, we don't know the contents. Plus even if we knew the contents (we kind of do), we don't know the exact width of the characters inside!
+
+### Approach 4: JavaScript
+
+There is prior art for this. jQuery plugins like [textFix](https://github.com/STRML/textFit) solve this problem by iteratively updating the font size of the element until it fits perfectly. The downside is, you're using JavaScript, in a loop, to measure and set font size. Performance-conscious developers hate this. The upside is, you're using regular DOM text elements! The scaling will be correct (modern browsers support sub-pixel scaling just fine), the text is selectable, announce-able, the whole deal. Spoiler: this is what we went with.
+
+### Bonus Half Approch: `canvas`
+
+There's a way (I hear) to render text to a canvas and use `measureText` to get the dimensions, but I don't have a lot of `canvas` experience, I don't know what the A11y implications are, I'm loath to jump out of React too much, etc.
```

Comment:
There aren't really any a11y implications, because you just use the canvas to measure, but then still render to and actual DOM node. This way you just avoid the cost of getBoundingClientRect and style trashing. I would need to try this again, but afaik the only downside is that you need to ensure that everything is correctly scaled by the device pixel ratio. In many cases, sites will have a single canvas that is hidden which can be referenced by all of the components so that you dont need to deal with creating a canvas element per component

Reply:
Roger, expanded this section with details, thanks!